text/0938-strict-es-module-support.md

```diff
@@ -0,0 +1,247 @@
+      ```
+ - Default module exports will no longer get automatically converted to namespace exports.
+ - Importing (including re-exporting) a non-existent name becomes an early error.
+ - Attempting to mutate a module from the outside throws an exception.
+ - `importSync` from `'@embroider/macros'` continues to work but it no longer guarantees lazy-evaluation (since lazy-and-synchronous is impossible under strict ES modules). 
+
+### Implication: No touching Ember from script context
+
+Today, an addon can use the `app.import` API in ember-cli to inject a script into the build, and that script can access parts of Ember via `require`. That behavior is intentionally made impossible by this change.
+
+Addon authors are encouraged to offer explicit module-based API instead, where apps are expected to import your special feature before importing anything else from Ember that you were trying to influence.
+
+### Implication: eager importSync
+
+The `importSync` macro:
+
+```js
+import { importSync } from '@embroider/macros'
+```
+
+exists primarily as a way to express conditional inclusion in places where Ember doesn't offer the ability to absorb asynchrony.
+
+This continues to work under strict-es-modules, but we can no longer guarantee *lazy evaluation* of the module that you `importSync`. 
+
+We still guarantee **branch elimination in production builds**, so if you say:
+
+```js
+import { importSync, macroCondition, getOwnConfig } from '@embroider/macros';
+
+let implementation;
+if (macroCondition(getOwnConfig().SOME_FLAG)) {
+  implementation = importSync('./new-one');
+} else {
+  implementation = importSync('./old-one');
+}
+```
+
+we continue to guarantee that in production builds your bundle will only ever contain one of those implementations and not both. But in development builds, both may be present *and evaluated* in your application.
+
+And if you don't do branch elimination at all:
+
+```js
+import { importSync } from '@embroider/macros';
+
+class Example extends Component {
+  onClick = () => {
+    let util = importSync('./some-utility').default;
+    util();
+  }
+}
+```
+
+your dependency is effectively static, as if you had written:
+
+```js
+import * as _some_utility  from './some-utility';
+
+class Example extends Component {
+  onClick = () => {
+    let util = _some_utility.default;
+    util();
+  }
+}
+```
+
+This can still be useful in the case of template-string-literal `importSync`:
+
+```js
+import { importSync } from '@embroider/macros';
+
+class Example extends Component {
+  onClick = () => {
+    let widget = importSync(`./widgets/${this.which}`);
+    widget.doTheThing();
+  }
+}
+```
+
+because it will automatically build into your app all the possible matches, when you want that.
+
+## Recommendations: Replacing `require` in Addons
+
+- If you're trying to import Ember-provided API that is only available on some Ember versions, use `dependencySatisfies` to guard the `import()` or `importSync()`.
```

Comment:
should caveats be added here that folks need to get their dep graph under control in some scenarios?

Reply:
Yeah, I said that explicitly in the `require.has` section but not this section.